codec-http2/src/main/java/io/netty/handler/codec/http2/BufferingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import static io.netty.handler.codec.http2.Http2CodecUtil.SMALLEST_MAX_CONCURRENT_STREAMS;
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.TreeMap;
+
+/**
+ * Implementation of a {@link Http2ConnectionEncoder} that dispatches all method call to another
+ * {@link Http2ConnectionEncoder}, until {@code SETTINGS_MAX_CONCURRENT_STREAMS} is reached.
+ * <p/>
+ * <p>When this limit is hit, instead of rejecting any new streams this implementation buffers newly
+ * created streams and their corresponding frames. Once an active stream gets closed or the maximum
+ * number of concurrent streams is increased, this encoder will automatically try to empty its
+ * buffer and create as many new streams as possible.
+ * <p/>
+ * <p>This implementation makes the buffering mostly transparent and can be used as a drop in
+ * replacement for {@link io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder}.
+ */
+public class BufferingHttp2ConnectionEncoder extends DecoratingHttp2ConnectionEncoder {
+
+    /**
+     * Thrown by {@link BufferingHttp2ConnectionEncoder} if buffered streams are terminated due to
+     * receipt of a {@code GOAWAY}.
+     */
+    public static final class GoAwayClosedStreamException extends Exception {
```

Comment:
@buchgr Failure requires some sort of exception, so I'm not sure this is really that bad of a solution.  We could just fail with a plain vanilla `StreamException` with error `STREAM_CLOSED`.  I actually this would probably be fine for gRPC.